### PR TITLE
Implement model exporting

### PR DESCRIPTION
**Pull request by Cwiiis**

### CODE REVIEW

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -532,26 +540,28 @@
     "\n",
     "The next line of the funtion `BiRNN` does a bit more data preparation.\n",
     "```python\n",
-    "    # Split data because rnn cell needs a list of inputs for the RNN inner loop\n",
-    "    layer_3 = tf.split(0, n_steps, layer_3)\n",
+    "    # Reshape data because rnn cell needs a list of inputs for the BRNN inner loop\n",
+    "    layer_3 = tf.reshape(layer_3, [-1, batch_size, n_hidden_3])\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Whoops, hard-coding the batch_size here - just needs to take it from the input instead, fixed locally.

File: DeepSpeech.ipynb [hunk 2]
```diff
@@ -1337,6 +1347,55 @@
...
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "with tf.device('/cpu:0'):\n",
+    "    model_exporter = exporter.Exporter(saver)\n",
+    "    \n",
+    "    # Run inference\n",
+    "    input = tf.placeholder(tf.float32, [None, None, 2*n_input*n_context])\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: The shape here should be [None, None, n_input + 2*n_input*n_context].

File: DeepSpeech.ipynb [hunk 3]
```diff
@@ -447,18 +455,20 @@
     "    # Backward direction cell\n",
     "    lstm_bw_cell = tf.nn.rnn_cell.BasicLSTMCell(n_cell_dim, forget_bias=1.0)\n",
     "    \n",
-    "    # Split data because rnn cell needs a list of inputs for the BRNN inner loop\n",
-    "    layer_3 = tf.split(0, n_steps, layer_3)\n",
-    "    \n",
+    "    # Reshape data because rnn cell needs a list of inputs for the BRNN inner loop\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This comment is now out of date with the code.

File: DeepSpeech.ipynb [hunk 4]
```diff
@@ -532,26 +544,28 @@
     "\n",
     "The next line of the funtion `BiRNN` does a bit more data preparation.\n",
     "```python\n",
-    "    # Split data because rnn cell needs a list of inputs for the RNN inner loop\n",
-    "    layer_3 = tf.split(0, n_steps, layer_3)\n",
+    "    # Reshape data because rnn cell needs a list of inputs for the BRNN inner loop\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This comment is now out of date with the code.

File: DeepSpeech.ipynb [hunk 5]
```diff
@@ -532,26 +544,28 @@
...
+    "    layer_3 = tf.reshape(layer_3, [-1, batch_x_shape[0], n_hidden_3])\n",
     "```\n",
-    "It splits `layer_3` in to `n_steps` tensors along dimension `0` as the LSTM BRNN expects its input to be of shape `n_steps *[batch_size, 2*n_cell_dim]`.\n",
+    "It reshapes `layer_3` in to `n_steps` tensors along dimension `0` as the LSTM BRNN expects its input to be of shape `[n_steps, batch_size, 2*n_cell_dim]`.\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This comment is now out of date with the code.

File: DeepSpeech.ipynb [hunk 6]
```diff
@@ -663,14 +677,14 @@
...
-    "    batch_seq_len = batch_x.shape[0] * [n_steps]\n",
+    "    batch_seq_len = batch_x.shape[0] * [batch_x.shape[1]]\n",
     "```\n",
     "creates `batch_seq_len` a list of the lengths of the sequences in `batch_x`. (As the sequences are zero padded to the same length, the list contains the value `n_steps` a total of `batch_x.shape[0]` times.)\n",
     "\n",
     "The next line\n",
     "```python\n",
     "    # Calculate the logits from the BiRNN\n",
-    "    logits = BiRNN(batch_x, n_steps)\n",
+    "    logits = BiRNN(batch_x)\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This comment is now out of date with the code.

File: DeepSpeech.ipynb [hunk 7]
```diff
@@ -615,13 +629,13 @@
    "source": [
     "def calculate_accuracy_and_loss(batch_set):\n",
     "    # Obtain the next batch of data\n",
-    "    batch_x, batch_y, n_steps = batch_set.next_batch()\n",
+    "    batch_x, batch_y = batch_set.next_batch()\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This code change is out of sync with the markup commenting on it below.

File: DeepSpeech.ipynb [hunk 8]
```diff
@@ -1337,6 +1351,55 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "with tf.device('/cpu:0'):\n",
+    "    model_exporter = exporter.Exporter(saver)\n",
+    "    \n",
+    "    # Run inference\n",
+    "    input = tf.placeholder(tf.float32, [None, None, n_input + 2*n_input*n_context])\n",
+    "    \n",
+    "    n_items  = tf.slice(tf.shape(input), [0], [1])\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Maybe add a comment on what these three lines that derive `seq_length` do.

File: DeepSpeech.ipynb [hunk 9]
```diff
@@ -1337,6 +1351,55 @@
...
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "with tf.device('/cpu:0'):\n",
+    "    model_exporter = exporter.Exporter(saver)\n",
+    "    \n",
+    "    # Run inference\n",
+    "    input = tf.placeholder(tf.float32, [None, None, n_input + 2*n_input*n_context])\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Could you choose another variable name, there's a builtin function named `input`. From `input`'s help page
```
Help on built-in function input in module __builtin__:

input(...)
    input([prompt]) -> value
    
    Equivalent to eval(raw_input(prompt)).
```

File: DeepSpeech.ipynb [hunk 10]
```diff
@@ -1337,6 +1351,55 @@
...
+    "    \n",
+    "    n_items  = tf.slice(tf.shape(input), [0], [1])\n",
+    "    n_steps = tf.slice(tf.shape(input), [1], [1])\n",
+    "    seq_length = tf.tile(n_steps, n_items)\n",
+    "    logits = BiRNN(input, tf.to_int64(seq_length))\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Maybe here and below can you add a few more comments on what's going on.

File: DeepSpeech.ipynb [hunk 11]
```diff
@@ -1337,6 +1351,55 @@
...
+    "    \n",
+    "    logits = tf.transpose(logits, [1, 0, 2])\n",
+    "    decoded, _ = ctc_ops.ctc_beam_search_decoder(logits, seq_length)\n",
+    "    decoded = tf.sparse_tensor_to_dense(decoded[0])\n",
+    "    \n",
+    "    # TODO: Transform output to a string\n",
+    "    \n",
+    "    # Restore variables from training checkpoint\n",
+    "    checkpoint = tf.train.get_checkpoint_state(checkpoint_dir)\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: I guess it doesn't matter now but we should think about which checkpoint we a recovering here. If we use the validation set to do early stopping the latest checkpoint file will not be the one we want to use.

File: DeepSpeech.ipynb [hunk 12]
```diff
@@ -1337,6 +1351,55 @@
...
+    "    input = tf.placeholder(tf.float32, [None, None, n_input + 2*n_input*n_context])\n",
+    "    \n",
+    "    n_items  = tf.slice(tf.shape(input), [0], [1])\n",
+    "    n_steps = tf.slice(tf.shape(input), [1], [1])\n",
+    "    seq_length = tf.tile(n_steps, n_items)\n",
+    "    logits = BiRNN(input, tf.to_int64(seq_length))\n",
+    "    \n",
+    "    logits = tf.transpose(logits, [1, 0, 2])\n",
+    "    decoded, _ = ctc_ops.ctc_beam_search_decoder(logits, seq_length)\n",
+    "    decoded = tf.sparse_tensor_to_dense(decoded[0])\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: What might be nice to do, and does not look like it would be too hard, is to output multiple results. Something like
```python
decoded = [tf.sparse_tensor_to_dense(d) for d in decoded]
```
It's pretty standard for STT engines to do this and it lets downstream code with domain knowledge to disambiguate between the various results.

File: DeepSpeech.ipynb [hunk 13]
```diff
@@ -84,14 +84,16 @@
...
     "import tensorflow as tf\n",
     "from util.log import merge_logs\n",
     "from util.gpu import get_available_gpus\n",
     "from util.text import sparse_tensor_value_to_texts, wers\n",
     "from tensorflow.python.ops import ctc_ops\n",
+    "from tensorflow.contrib.session_bundle import exporter\n",
     "\n",
     "ds_importer = os.environ.get('ds_importer', 'ted_lium')\n",
-    "ds_dataset_path = os.environ.get('ds_dataset_path', './data/ted')\n",
+    "ds_dataset_path = os.environ.get('ds_dataset_path', os.path.join('./data', ds_importer))\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: It's unrelated, was just annoyed at how many times I got caught out changing ds_importer and forgetting to change ds_dataset_path. I've also got the changeset that makes us use XDG_DATA_HOME instead of /tmp in here too, shall I split them out into a separate PR?

File: DeepSpeech.ipynb (hunk 12 above)
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I've done what you suggested in anticipation of us picking a value for `top_path` later.

**Comment by kdavis-mozilla** [comment]
@Cwiiis LGTM FYI @tilmankamp also has a pull request in the queue to be merged. So, there may be some merge conflicts that need to be worked out depending on who goes first.

**Comment by kdavis-mozilla** [comment]
@Cwiiis @tilmankamp I'll let you two figure out who merges first and how the conflicts are resolved.
